Mellanox/ompi
# Reworked Mellanox OpenMPI CI with Azure Pipelines service v1.0

# Mellanox OpenMPI CI
## Scope
[Mellanox](https://www.mellanox.com/) OpenMPI CI is intended to verify OpenMPI with recent Mellanox SW components ([Mellanox OFED](https://www.mellanox.com/page/products_dyn?product_family=26), [UCX](https://www.mellanox.com/page/products_dyn?product_family=281&mtag=ucx) and other [HPC-X](https://www.mellanox.com/page/products_dyn?product_family=189&mtag=hpc-x) components) in the Mellanox lab environment.

CI is managed by [Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/?view=azure-devops) service.

Mellanox OpenMPI CI includes:
* OpenMPI building with internal stable engineering versions of UCX and HCOLL. The building is run in Docker-based environment.
* Sanity functional testing.
## How to Run CI
Mellanox OpenMPI CI is triggered upon the following events:
* Push a commit into the master branch. CI status and log files are available on the Azure DevOps server.
* Create a pull request (PR). CI status is visible in the PR status. CI is restarted automatically upon each new commit within the PR. CI status and log files are also available on the Azure DevOps server.
* Trigger CI with special PR comments (for example, `/azp run`). Comment triggers are available only if the comment author has write permission to the PR target repo. Detailed information about comment triggers is available in the official Azure DevOps [documentation](https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#comment-triggers).

# TODO
## Future Improvements
1. Run several executors on the agent
2. Align versions of OS/MLNX_OFED with HPC-X ones
3. Need to reduce CI time (faster CPU, SSD, etc.)
4. Adapt taskset usage in scripts for AzureCI (if applicable - if several executors are allowed on an agent)
5. Use agent hosts with NUMA enabled
6. Fix code inspection warnings (non-critical) in the shell scripts
7. Use release HPC-X instead of module
8. Use Dockerhub instead of Mellanox Harbor
9. Rework ompi_test.sh (split build and test)

## Verification scenarios
1. Verify `/azp run` in PR comments by non-Azure users
UPD: comment triggers are available only if the comment author has write permission to the PR target repo (so for the official OMPI repo most developers won't be able to use comment triggers). See https://github.com/Mellanox/ompi/pull/7
2. Verify non-master PRs
UPD: passed (see https://github.com/Mellanox/ompi/pull/8)
3. Verify master per-commit CI
UPD: passed
4. Verify that PR ompi sources are correctly used within CI (check git source tree)
UPD: passed.
5. Check CI triggering for PRs which exist before PR enabling
UPD: AzureCI isn't triggered (azure-pipelines.yml is absent)
6. Verify parallel jobs.
UPD: passed (see https://github.com/Mellanox/ompi/pull/7)
7. Verify that CI is triggered for PRs from non-Azure users
UPD: passed (see https://github.com/Mellanox/ompi/pull/7)
8. Non-master PR within a fork.
UPD: CI isn't triggered (expected) - see https://github.com/itemko-tester/ompi/pull/1
